### PR TITLE
lib.packagesFromDirectoryRecursive: init

### PR DESCRIPTION
**Pull request by 9999years**
## Description of changes

Adds a `packagesFromDirectoryRecursive` helper, which combines `builtins.readDir` and `callPackage` to transform a directory of `.nix` package files into a matching nested attribute set of derivations.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by infinisil — DISMISSED** [review]
Can you split out the documentation updates into a separate PR?

Regarding `packagesFromDirectory`, I'd rather make this be unified with `pkgs/by-name` than a separate convention. The 2-letter prefix of `pkgs/by-name` should definitely be optional, and the ability to use files or directories only could also be optional.

**Comment by 9999years** [comment]
Split the documentation updates into #270696.

**Comment by 9999years** [comment]
OK, I've unified `lib.packagesFromDirectory` with the `pkgs/by-name` overlay. The `pkgs/by-name` tree is already set up in the same manner `packagesFromDirectory` expects, so I didn't need to make as many changes as anticipated (there's no extra directories without a `package.nix` for `packagesFromDirectory` to recurse into).

I've also reimplemented `packagesFromDirectory` in terms of a new helper, `lib.packageFilesFromDirectory`, which just returns a set of package paths and skips invoking `callPackage`. This also lets us hoist the directory traversal portion of the `pkgs/by-name` out of the overlay function itself, to save evaluation time if the overlay is applied multiple times.

**Comment by infinisil** [comment]
Thinking about it again, this does sound distinct enough that maybe it shouldn't be shared with `pkgs/by-name`. Especially since the latter comes with its own CI checking tool (`pkgs/tests/nixpkgs-check-by-name`) and [an RFC](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md).

I'm afraid I don't have a lot of capacity to review this right now though, I hope others can help out with this.

**Comment by 9999years** [comment]
But this is totally compatible with `pkgs/by-name` and the `pkgs/tests/nixpkgs-check-by-name` tool?

**Comment by infinisil** [comment]
The function seems to support a _superset_ of `pkgs/by-name` functionality. Using files instead of directories is only supported by function, not `pkgs/by-name`. Same for the recursive sets. Notably the additional checking of `pkgs/by-name` is only implemented in CI. So people could e.g. create `pkgs/by-name/foo/foo.nix` locally, and it works just fine locally, but then CI complains. Comparatively before this PR it just didn't work.

Furthermore, as soon as you have a package set there's the requirement of being able to override it. If you use `lib.makeScope` this gets handled somewhat decently at least for one level, but it's going to get messy for multiple levels. This really needs a lot of thought, and I don't have the capacity for this right now.

File: lib/filesystem.nix
Comment (Gabriella439) [inline lib/filesystem.nix:323]: The main thing I'll suggest is to make this take an attribute set instead of position arguments:

```lib
{ callPackage, directory }:
```

… that way if we need to add new optional arguments (with sensible defaults) we don't have to break backwards-compatibility.  This would also be consistent with how `pkgs.haskell.lib.packagesFromDirectory` works (it also takes an attribute set)

**Review by Gabriella439 — COMMENTED** [review]
I'm filling in to review since @infinisil indicated that he doesn't have time to give this a full review.

I think this is still okay to merge with just one small change.  In particular, I'm not too concerned that the `pkgs/by-name` function might (hypothetically) in the future have different requirements than `lib.packagesFromDirectory`.  If we ever get to that point it won't be hard to split the two implementations, but as far as I'm concerned the fact that this can be reused to power`pkgs/by-name` functionality in its present state is a feature.

File: lib/filesystem.nix
Comment (Gabriella439) [reply lib/filesystem.nix:323]: Oh, and also, have the attribute set ignore extra arguments (for forward compatibility), also the same as `pkgs.haskell.lib.packagesFromDirectory`:

```nix
{ callPackage, directory, ... }:
```

Comment (9999years) [reply lib/filesystem.nix:323]: Done.

**Review by Gabriella439 — APPROVED** [review]

File: doc/using-nixpkgs.md
Comment (infinisil) [inline doc/using-nixpkgs.md:8]: This change should be discussed and reviewed separately in https://github.com/NixOS/nixpkgs/pull/270696

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:49]: Don't run code formatters unless it's really necessary (which I don't think is the case here).

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:257]:
```suggestion
                  throw
```

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:263]: Error messages in `lib` should look more like this:
```suggestion
                      lib.filesystem.packagesFromDirectory: Unsupported file type ${type} on path ${toString path}
```

Also very important to use `toString path`. Just using `"${path}"` imports the entire path into the store!

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:243]: Use `hasSuffix ".nix"` and `removeSuffix ".nix"` instead. The `replaceStrings` is slightly incorrect because it can also replace parts of the beginning of the string.

File: pkgs/test/packages-from-directory/default.nix
Comment (infinisil) [inline pkgs/test/packages-from-directory/default.nix:1]: Tests for `lib` should be in `lib/tests` (or you can also create `lib/filesystem/tests`, see also https://github.com/NixOS/nixpkgs/tree/master/lib#running-tests)

File: pkgs/top-level/by-name-overlay.nix
Comment (infinisil) [inline pkgs/top-level/by-name-overlay.nix:28]: As already mentioned before, I don't think anything should be changed in this file. Yes I did ask to make it be shared with `pkgs/by-name` originally, but I don't think that should be done anymore. It's distinctly different that just having `packageFilesFromDirectory` for this very use case doesn't make any sense, especially when half of its functionality can't be used here anyways.

And considering that, we don't need this separate function.

**Review by infinisil — CHANGES_REQUESTED** [review]
Sorry but this PR is definitely not mergeable as is. Saying that I don't have time for a review doesn't mean that my feedback shouldn't be addressed. Given that I am the de-facto maintainer for a lot of the files you touch here, I _should_ definitely review this, so here I go. It's not a super detailed review, but I can do another pass once the points are addressed.

Also take a look at https://github.com/NixOS/nixpkgs/pull/272083 where me and the other effective `lib` maintainer roberth discuss `lib` contribution guidelines.

File: doc/using-nixpkgs.md
Comment (9999years) [reply doc/using-nixpkgs.md:8]: How do I link to nix.dev documentation here? I assume you want these docs on nix.dev.

**Review by tomberek — COMMENTED** [review]
I'm not sure this callPackage will do the right thing with packages that depend on each other. I can imagine using this via an identity function or an import and then using a standard callPackage. 

That being said, this is a good addition into our standard library, often needed.

**Comment by 9999years** [comment]
@tomberek Yeah, the documentation I was asked to remove made this clearer, but `packagesFromDirectory` is most helpful in combination with `lib.makeScope`.

File: doc/using-nixpkgs.md
Comment (infinisil) [reply doc/using-nixpkgs.md]: Once merged you can just link to the nix.dev page directly

**Comment by RaitoBezarius** [comment]
I'm not a maintainer of lib per se but this seems to be a very useful addition and there's many out of tree code that reinvents the same thing all the time, it would be tidy if we could just reuse nixpkgs for that directly.

File: lib/tests/packages-from-directory/default.nix
Comment (infinisil) [inline lib/tests/packages-from-directory/default.nix:15]: No need to involve derivations. The functions works for any Nix value, so you could just use the string directly.

```suggestion
    name
```

(and then some things may simplify)

File: lib/tests/packages-from-directory/default.nix
Comment (infinisil) [inline lib/tests/packages-from-directory/default.nix:10]: No need to involve derivations. The functions works for any Nix values. This should work:

```nix
packagesFromDirectory {
  packageFromFile = import;
```

And then just put strings directly into the package files.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix:240]: Just about everyone will want to use `callPackage` here, I don't think users should have to write a `path: pkgs.callPackage path {}` wrapper to use this. For special cases it's just as easy to go the other way (`path: _: import path`).

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix:214]: Is this a nixpkgs restriction elsewhere? I think it would require `package/package.nix` with this code, but it should be possible.

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix]: Done, chose `packagesFromDirectoryRecursive` to match (e.g.) `filterAttrsRecursive`.

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix]: But also I could see adding another argument to provide overrides, maybe bundling this up with `lib.makeScope`:

```nix
packagesFromDirectoryRecursive {
  overrides = self: {
    packageName = /* overrides set */;
  };
}
```

Comment (tomberek) [reply lib/filesystem.nix]: Yes, an ergonomic variant of makeScope is good. I'll try to elaborate later.

File: lib/filesystem.nix
Comment (infinisil) [reply lib/filesystem.nix:222]: Ah yeah that works, though it can easily trip up people when they start out with
```
nested
├── a.nix
└── b.nix
```

But then add a `package.nix`, and suddenly the output structure changes, ignoring the other files.

However, if it's not a nested package set, `package.nix` does actually work just fine to declare `package` in the current implementation! It might make sense to flip around the recursion to make this consistent, such that if the `directory` itself contains a `package.nix`, it turns into a single package, not a set.

File: lib/filesystem.nix
Comment (infinisil) [reply lib/filesystem.nix]: Having talked on Matrix a bit, I do agree that `callPackage` would be nice to have just because it's so well-known, even if it doesn't really make sense for the function as is. So I think having `callPackage` as an argument is fine.

I don't really want to block on this, because this is a useful function, even if it's not perfect.

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix:222]: > It might make sense to flip around the recursion to make this consistent, such that if the `directory` itself contains a `package.nix`, it turns into a single package, not a set.

I think this would have to be a special case because then `packagesFromDirectoryRecursive` wouldn't return an attribute set.

Comment (9999years) [reply lib/filesystem.nix:225]: Actually I managed to make it work rather nicely. PTAL

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix]: I'll mark this as resolved for now and work on adding `packageFromPath` once we get this merged.

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:223]: Was a bit unclear what "or" means here. How about
```suggestion
        As a result, directories with no `.nix` files (including empty directories)
```

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:251]:
```suggestion
          (Path -> a) -> AttrSet -> a
```

Will always be called with a `Path`

File: lib/filesystem.nix
Comment (infinisil) [inline lib/filesystem.nix:296]: I wonder if this is actually worth having now, because it only occurs when you have a `.nix` entry that's neither a directory nor a regular file. All that's left is either symlinks, which would be okay to use, or unknown filetypes such as `mkfifo`s, in which case the user probably knows what they're doing (and one can totally read Nix from a `mkfifo`!).

This then allows some code simplification, also making it mirror the above `directoryEntryIsPackage`:

```suggestion
        if type == "directory" then
          {
            name = basename;
            value = packagesFromDirectory path;
          }
        else
          {
            name = removeSuffix ".nix" basename;
            value = callPackage path { };
          };
```

(also I'm suggesting a different formatting here according to https://github.com/NixOS/rfcs/pull/166 :smile:)

Then you could even add a symlink to the test case.

File: lib/tests/misc.nix
Comment (infinisil) [inline lib/tests/misc.nix:2019]: Since the overrides are always empty for now, you could also do

```suggestion
      callPackage = path: overrides: import path;
```

Not a blocker though.

**Review by infinisil — COMMENTED** [review]
Just minor comments, looking pretty good now!

File: lib/tests/misc.nix
Comment (9999years) [reply lib/tests/misc.nix:2019]: Yeah, I just wanted to keep it in "the spirit" of `callPackage`.

File: lib/filesystem.nix
Comment (9999years) [reply lib/filesystem.nix:296]: I tried this, but because Nix can't differentiate between symlinks to files and symlinks to directories, the behavior was pretty confusing -- symlinks to individual nix files worked fine, but directories didn't show up at all.

Comment (infinisil) [reply lib/filesystem.nix:296]: Ah yeah that's a good point!

**Review by infinisil — APPROVED** [review]

**Comment by infinisil** [comment]
Not gonna bother waiting for CI, it's too slow.. I tested it locally, all good :+1:

**Comment by philiptaron** [comment]
I really love this function. Thanks for adding it @9999years. 🙇🏻

File: lib/filesystem.nix
Comment (Gabriella439) [reply lib/filesystem.nix]: I'd be fine removing the `...`.  I'm not *too* attached to it.

Comment (infinisil) [reply lib/filesystem.nix]: +1 for dropping `...`!

**Pull request merged**
